contracts/dao/neutron-flashloans/src/contract.rs

```diff
@@ -0,0 +1,403 @@
+                    // We sent the return address to the source address
+                    return_address: config.source,
+                    loan_amount: active_loan.amount,
+                    fee: active_loan.fee,
+                })
+                .unwrap(),
+                funds: vec![],
+            });
+            // We use the reply_on_success strategy (we want the transaction to be simply reverted
+            // in case of an error).
+            Ok(Response::new()
+                .add_submessage(SubMsg::reply_on_success(msg, BORROWER_HANDLER_REPLY_ID))
+                .add_attribute("action", "reply_authz_bank_send"))
+        }
+        BORROWER_HANDLER_REPLY_ID => {
+            // If we are here, the borrower smart contract has successfully executed the
+            // ProcessLoan message, and probably returned the funds to our contract together with
+            // the additional fee.
+
+            let config: Config = CONFIG.load(deps.storage)?;
+            let active_loan = must_get_active_loan(&deps)?;
+
+            // Check that the borrower returned the loan and paid the fee
+            check_expected_balances(deps.as_ref(), config.source, active_loan.expected_balances)?;
+
+            // Set the active loan to None, thus making ourselves ready for the next loan
+            ACTIVE_LOAN.save(deps.storage, &None)?;
+
+            Ok(Response::new().add_attribute("action", "reply_borrower_handler"))
+        }
+        _ => Err(ContractError::UnknownReplyID { id: msg.id }),
+    }
+}
+
+fn check_expected_balances(
+    deps: Deps,
+    source: Addr,
+    expected_balances: Vec<Coin>,
+) -> Result<(), ContractError> {
+    // Prepare the query
+    let all_balances_query = BankQuery::AllBalances {
+        address: source.to_string(),
+    };
+    // Get the response (all balances)
+    let all_balances_response: AllBalanceResponse =
+        deps.querier.query(&all_balances_query.into())?;
+
+    // For each of the expected coin balances, check that the current balance of the source
+    // matches the expectations. We require **exactly** the expected amount (loan amount + fee)
+    // to be transferred back to the source, not more, not less.
+    for expected_coin in expected_balances {
+        let maybe_actual_balance = all_balances_response
+            .amount
+            .iter()
+            .find(|x| x.denom == expected_coin.denom && expected_coin.amount.eq(&x.amount));
+
+        if maybe_actual_balance.is_none() {
+            return Err(ContractError::IncorrectPayback {});
+        }
+    }
+
+    Ok(())
+}
+
+// Loads the information about the current loan, returns an error if the information
+// is missing.
+fn must_get_active_loan(deps: &DepsMut) -> Result<ActiveLoan, ContractError> {
+    let active_loan = ACTIVE_LOAN.load(deps.storage)?;
+    if active_loan.is_none() {
+        return Err(ContractError::UnexpectedState {});
+    }
+
+    Ok(active_loan.unwrap().clone())
```

Comment:
Unnecessary `.clone()` call

```suggestion
    Ok(active_loan.unwrap())
```